ui/vty-lib/src/Ariadne/UI/Vty/Scrolling.hs

```diff
@@ -84,3 +87,61 @@ fixedViewport name vpType p = case vpType of
       B.Vertical -> B.hLimit (result ^. B.imageL & V.imageWidth)
       B.Horizontal -> B.vLimit (result ^. B.imageL & V.imageHeight)
       B.Both -> identity
+
+-- | Creates a viewport with a scrollbar on the right, bottom or both sides.
+-- Note: when there is no need the scrollbar does not get displayed, but
+-- it will still give one column/row less to the sub-widget to use.
+viewportWithScrollBar
+    :: (Ord n, Show n)
+    => n                -- ^ The name of the viewport (must be unique)
+    -> B.ViewportType   -- ^ The type of the viewport (scrolling direction)
+    -> B.Widget n       -- ^ The widget to be rendered in the viewport
+    -> B.Widget n
+viewportWithScrollBar name vpType p = B.Widget B.Greedy B.Greedy $ do
+    -- first get the current context
+    c <- B.getContext
+    -- adjust it for the inner widget
+    let unrestricted = 100000
+        releasing = case vpType of
+            B.Vertical   -> set B.availHeightL unrestricted
+                              . over B.availWidthL (subtract 1)
+            B.Horizontal -> set B.availWidthL unrestricted
+                              . over B.availHeightL (subtract 1)
+            B.Both       -> over B.availHeightL (subtract 1)
+                              . over B.availWidthL (subtract 1)
+    -- render the inner widget
+    result <- R.withReaderT releasing $ B.render p
+    -- add the scrollbar to the result
+    mvp <- B.unsafeLookupViewport name
+    let vpWidget = B.viewport name vpType $
+            B.Widget (B.hSize p) (B.vSize p) (return result)
+
+        resultImg = B.image result
+        (availH, availW) = case vpType of
+            B.Both -> (B.availHeight c - 1, B.availWidth c - 1)
+            _ -> (B.availHeight c, B.availWidth c) 
+    
+        enoughHeight = V.imageHeight resultImg <= availH
+        availHeight = fromIntegral availH
+        resultHeight = fromIntegral $ V.imageHeight resultImg
+        vpPadTop = fromIntegral $ maybe 0 (view B.vpTop) mvp
+        vRatio = availHeight / resultHeight :: Double
+        vPad = round $ vRatio * vpPadTop 
+        vSize = max 1 . round $ vRatio * availHeight
+        vBar = B.vLimit (vPad + vSize) $ B.padTop (B.Pad vPad) BR.vBorder
+        addVBar widg = if enoughHeight then widg else B.hBox [widg, vBar]
+
+        enoughWidth = V.imageWidth resultImg <= availW
+        availWidth = fromIntegral availW
+        resultWidth = fromIntegral $ V.imageWidth resultImg
+        vpPadLeft = fromIntegral $ maybe 0 (view B.vpLeft) mvp
+        hRatio = availWidth / resultWidth :: Double
+        hPad = round $ hRatio * vpPadLeft
+        hSize = max 1 . round $ hRatio * availWidth
+        hBar = B.hLimit (hPad + hSize) $ B.padLeft (B.Pad hPad) BR.hBorder
+        addHBar widg = if enoughWidth then widg else B.vBox [widg, hBar]
```

Comment:
I would create one function like `availDim -> imageDim -> (barSize, barPad)` and use it for both horizontal or vertical bars